library/src/main/java/com/google/android/exoplayer/audio/AudioTrack.java

```diff
@@ -414,10 +412,15 @@ public int handleBuffer(ByteBuffer buffer, int offset, int size, long presentati
       bytesWritten = writeNonBlockingV21(audioTrack, buffer, temporaryBufferSize);
     }
 
-    temporaryBufferSize -= bytesWritten;
-    submittedBytes += bytesWritten;
-    if (temporaryBufferSize == 0) {
-      result |= RESULT_BUFFER_CONSUMED;
+    if (bytesWritten < 0) {
+        Log.w(TAG, "AudioTrack.write returned error code: " + bytesWritten);
```

Comment:
The idea was to at least get one warning. The previous code did not log errors in the 'SDK >= 21' case. So it took me some time to find the root cause. But the failing was permanent indeed, I have never been able to recover such an error so failing playback would be good as well